Use brand new PKGBUILD for Aseprite Package

ISSOtm opened the pull request:
This attempts to fix the old one's jank.

- Dependencies are pulled via the bundled script for more reliability
- Less static libs = yes
- `CMAKE_BUILD_TYPE` set to `None` to use Arch Linux's build flags
- Use bundled `gn` (Arch Linux's `gn` package appears to use an incompatible version. Whatta shame!)
- Use `libxxx` dependency specifications (where possible) to pin lib versions and avoid breakage
- `$pkgname` only used where it would help "forking" the PKGBUILD, i.e. not in source links, for example
- Python 2 not required to build
- Removed `conflicts` with other `aseprite` variants, as they will conflict if they properly `provides` and `conflicts` with `aseprite`

TODOs:
- [x] Not all dependencies checked out by `git-sync-deps` are necessary, since a lot of shared libs are used. Figure out a way to only sync what we need?
- [x] Make `libwebp` shared as well? Could use https://github.com/aseprite/aseprite/pull/2523 as a base.
- [x] `namcap` complains that `libexpat` and `libharfbuzz` are not used, despite `ldd`'s output. I'm a bit puzzled.
- [x] Is `icu` required at runtime? It's not linked to, but I \*think\* the headers might be required nonetheless. (Unfortunately, verifying this is a bit difficult.)
- [ ] Run tests? E.g. LAF's, except that they fail to build due to an obscure linker error (looks like some static lib requires a dynamic lib that doesn't get `-l`'d?)
- [x] Make sure we have feature parity with the current PKGBUILD, e.g. I noticed this one doesn't install icons in the same location
- [x] Did I miss any `makedepends`? Nothing direct for sure, but maybe some transitive dependencies (e.g. Python 2 was only required by a script indirectly invoked via `gn`)
- [x] Should we require `clang`? Skia's build instructions mention that performance may be significantly worse with non-Clang compilers
- [x] Do we really conflict with `skia-git`? If so, that should be reinstated in the `conflicts` array

(Aside: if you'd like, I can be a co-maintainer of the AUR package as well.)

ISSOtm committed:
Use brand new PKGBUILD instead

Attempting to fix the old one's jank

ISSOtm committed:
Install third-party licenses as well

ISSOtm committed:
Install .desktop file as well

ISSOtm committed:
Add missing build-time dependencies

ISSOtm commented:
Findings about build deps:
- `freeglut` appears to only be required by `imgui` examples (we don't build them) and `libwebp`'s `vwebp` (we don't build it either)
- I think the following transitive dependencies can be safely left implicit: `xorgproto` ← `libx11` ← `libxrender` ← [ `libxcursor` ; `libxrandr` ]. `libxrandr` is required by a few third-party files, but is not linked against in the final binary, so I'm not entirely sure.
- `nettle` is depended upon by `curl` and `libarchive`, but we use shared libs for both, so it can be removed
- `harfbuzz-icu` is required by Skia... unless using the bundled version. So we can also get rid of the dependency.

ISSOtm committed:
Install icons in the icon theme directories instead

Also remove redundant `$srcdir`s in `package()`

ISSOtm committed:
Put font license in the central directory as well

ISSOtm committed:
Bump pkgrel

It was bumped during my work

ImperatorStorm commented on PKGBUILD:
```diff
@@ -9,213 +9,106 @@
+        # Python 3-compliant version of the script
+        is_clang.py
+        # Based on https://patch-diff.githubusercontent.com/raw/aseprite/aseprite/pull/2535.patch
+        shared-libarchive.patch)
```
You need to include `is_clang.py` and `shared-libarchive.patch` in the repo. Currently, they aren't.
Probably will need to modify the `.gitignore` for this.

ISSOtm replied:
Right, that's my bad.

ImperatorStorm commented on PKGBUILD:
```diff
@@ -9,213 +9,106 @@
-    ..
+	echo Building Skia...
+	cd skia
+	local skiadir="$PWD"
```
https://wiki.archlinux.org/title/Arch_package_guidelines#Package_etiquette
should be `_skiadir`

ISSOtm replied:
AIUI, using `local` is fine, as such variables would not override anything outside of the function itself, and we don't use any `makepkg`-specific features inside the function that could be affected by our own.

ImperatorStorm replied:
Still, I'd rather be safe than sorry.

ISSOtm replied:
Ok, no harm in belt and braces I guess.

ImperatorStorm commented on PKGBUILD:
```diff
@@ -9,213 +9,106 @@
+pkgrel=4
 pkgdesc='Create animated sprites and pixel art'
 arch=('x86_64')
 url="http://www.aseprite.org/"
```
Mind bumping this to https?

ISSOtm replied:
The URL, you mean?

ImperatorStorm replied:
yes.

ISSOtm committed:
Ignore files in subdirectories

ISSOtm committed:
Add missing resource files

ISSOtm committed:
Use underscores for local variable names

https://wiki.archlinux.org/title/Arch_package_guidelines#Package_etiquette
https://github.com/ImperatorStorm/PKGBUILDs/pull/1#discussion_r777122781

ISSOtm committed:
Use HTTPS for upstream URL

ImperatorStorm commented:
Add yourself on as a Maintainer and I'll merge.

ISSOtm committed:
Add myself as co-maintainer

Thanks!

ISSOtm commented:
Thank you very much! Do we want to merge with the first post's TODOs still unresolved?

ImperatorStorm commented:
Now that you mention it,
> Not all dependencies checked out by git-sync-deps are necessary, since a lot of shared libs are used. Figure out a way to only sync what we need?

This should be done before merge
Everything else was either already fixed or is fine.

Perhaps patching [`DEPS`](https://github.com/aseprite/skia/blob/aseprite-m81/DEPS) to only grab the dependencies we need it to grab?

Alternatively, manually sync the deps?

ISSOtm commented:
I'm not sure which of these options is best. I'd rather sync them manually to avoid patching, but this might be cumbersome.

I'm also looking into dependencies again, especially `icu` which is not linked against despite `skia_use_system_icu=true`... tbh, it's quite hard to trace through all of the dependency chain.

ISSOtm commented:
Ok, so, passing `-v` to `ninja`, and then `grep -o -- '-I\s*../third_party/externals/\S*' | sort | uniq`, it follows that only `freetype` and `libgifcodec` are being used, at least like this.

More investigation is required, since I remember not being able to compile without at least `dng_sdk`. But at least we should be able to get somewhere.

ISSOtm commented:
Looking at `src/skia/gn/skia.gni`, it seems that the components that are being used are:

Component       | Used? | Reason
----------------|-------|--------
`expat`         | No    | Using system's
`fontconfig`    | No    | Using system's... I think?
`freetype`      | Yes   | Using system's `freetype2`, *but* the bundled `harfbuzz` requires `freetype` (!)
`harfbuzz`      | Yes   | Can't use system's due to an API change, apparently, so must use bundled
`gl`            | No    | ??
`icu`           | No    | Using system's
`libjpeg-turbo` | No    | Using system's
`libpng`        | No    | Using system's
`libwebp`       | No    | Using system's
`piex`          | No    | Must use bundled
`x11`           | No    | ??
`xps`           | No    | ??
`zlib`          | No    | Using system's
`dng_sdk`       | Yes   | By default if using `libjpeg-turbo` and `zlib`... huh.
`libgifcodec`   | Yes   | By default if not using `wuffs`
`sfntly`        | No    | Enabled by default since we use `icu`, but Aseprite's build instructions disable it... except in the Linux example??

This table needs more work, esp. figuring out which libs are required for what (mainly, what file formats does Skia need to load? What is handled directly by Aseprite?)
So consider the above a WIP. (Thankfully, we can already get rid of a *lot* of deps—many of them, such as `imgui`, are only used for building example programs, so we don't need them either.)

Looking at the various `BUILD.gn`, it seems that `is_official_build` sets `use_system_*=true` by default, so that would also simplify the `gn` config line.

I am also disabling `skia_enable_pdf` and `skia_enable_skottie`, as it seems that Aseprite uses neither. This should reduce the compile time.

ISSOtm committed:
Remove ICU dependency

We are using a statically-linked version instead, apparently

ISSOtm committed:
Avoid interactive prompts when patching fails

ISSOtm committed:
Clean up $srcdir

Extract Aseprite's sources into a separate directory
Configure Aseprite out-of-tree
Configure Skia in a fresh directory (apparently `gn` does not support out-of-tree)

ISSOtm committed:
Disable Skottie in Skia

This functionality doesn't appear to be used by Aseprite,
and excluding it should reduce build time

ISSOtm committed:
Only pull Skia dependencies that we need

This significantly reduces initial build time (from syncing large repos)
and storage use.
The only remaining dependency that has a chance to be axed is `dng_sdk`,
for which more investigation is needed.

ISSOtm committed:
Avoid printing redundant flags in `gn` args

Brainfart.
"static" flags like `is_debug` were re-printed for each "non-static"
(e.g. `skia_use_*`) variant.
Doing it this way also allows getting rid of weird quote shenanigans.

ISSOtm committed:
Simplify `gn` configure line

`is_official_build=true` sets Skia up to use system libs by default,
so only specify those that we *don't* want

ISSOtm committed:
Pull `gn` version ourselves

That way, we can additionally check its integrity via the SHA256

ISSOtm commented:
Ok, this should suffice for all the "cleanup" commits, and I've taken out 4 items from the OP's TODO list. I noticed that PDF support isn't actually disabled! Only some parts of it are, right now. I'll next look into axing that.

I'll check if e.g. disabling `libpng` causes Aseprite to fail linking, since I know for a fact that it expects to load and save PNG images. If it does, then we should be able to rely on disabling required features failing to build entirely; otherwise, we'd see missing features only have an impact at runtime.

Regardless, the bulk of the work should be done—I still have a lingering fear that I broke *something* with this amount of changes, but it compiles and runs fine for me—so this should be ready for review.

ImperatorStorm commented:
Workflow failed, missing `libglvnd` makedependency, possibly runtime dep.

```
[289/912] compile ../src/gpu/gl/glx/GrGLMakeNativeInterface_glx.cpp
FAILED: obj/src/gpu/gl/glx/gpu.GrGLMakeNativeInterface_glx.o 
c++ -MD -MF obj/src/gpu/gl/glx/gpu.GrGLMakeNativeInterface_glx.o.d -DNDEBUG -DSK_R32_SHIFT=16 -DSK_GAMMA_APPLY_TO_A8 -DSKIA_IMPLEMENTATION=1 -DSK_GL -I.. -fstrict-aliasing -fPIC -O3 -fdata-sections -ffunction-sections -Wno-unused-parameter -std=c++17 -fno-exceptions -fno-rtti -c ../src/gpu/gl/glx/GrGLMakeNativeInterface_glx.cpp -o obj/src/gpu/gl/glx/gpu.GrGLMakeNativeInterface_glx.o
../src/gpu/gl/glx/GrGLMakeNativeInterface_glx.cpp:15:10: fatal error: GL/glx.h: No such file or directory
   15 | #include <GL/glx.h>
      |          ^~~~~~~~~~
```

ImperatorStorm commented on PKGBUILD:
```diff
@@ -9,106 +10,54 @@
+         # Already required by Aseprite: libjpeg-turbo libpng zlib freetype2
+         # These two are only reported by Namcap, but don't seem to be direct dependencies?
+         libfontconfig.so libxcursor)
+makedepends=(# "Meta" dependencies
```
missing `libglvnd` in `makedepends`

ISSOtm replied:
Yeah, I noticed the run failed an got to work. I think the proper dependency is `libgl`, but [see my commit's message](https://github.com/ImperatorStorm/PKGBUILDs/pull/1/commits/60524054ec445a5a960436aa93ec8f9675162761) for why I am unsure about the runtime-ness of this dep.

Thanks for setting up the CI!

ImperatorStorm replied:
np.

ISSOtm committed:
Add forgotten `libgl` dependency

What's weird is that the lib seems not to be linked to dynamically,
but I can't see it not being required at run time. Right?

ISSOtm commented:
Something else: I'm noticing that the old PKGBUILD disabled the updater, but I currently am not (and thus it's enabled by default). Should we disable it?

ImperatorStorm commented:
Yes, updater's likely gonna be broken with this packaging.

ISSOtm committed:
Disable updater

https://github.com/ImperatorStorm/PKGBUILDs/pull/1#issuecomment-1003838784

ISSOtm commented:
New build fail, this time it's related to `libwebp`. I'll look at it tomorrow.

ImperatorStorm commented:
Alright.

ISSOtm commented:
Side note: I don't like that `libwebp` is being linked against dynamically by Skia, but not by the final executable. I'm a bit rusty on my static lib innards, but it seems possible that we'd be compiling Skia against the system headers, but linking against Aseprite's bundled lib.

ISSOtm commented:
Yeah, I am afraid that's what's happening. The fix seems to be making the same lib consistently shared or static in both Skia and Aseprite.

For `libwebp`, we should make it shared by Aseprite as well; for `freetype` and `harfbuzz`, we should make Skia point to Aseprite's bundled headers (or the opposite, whichever works). I think that should work.

ISSOtm committed:
Avoid creating symlinks to Skia deps in their dirs

Would occur if they were already symlinked

ImperatorStorm committed:
Maybe use system libwebp? Based on https://patch-diff.githubusercontent.com/raw/aseprite/aseprite/pull/2535.patch

ImperatorStorm commented:
pushed a commit based on https://patch-diff.githubusercontent.com/raw/aseprite/aseprite/pull/2523.patch, hopefully this fixes the libwebp issue.

Well, that's certainly an error

oh, its `USE_SHARED_WEBP` not `USE_SHARED_LIBWEBP`

ImperatorStorm committed:
Fix system `libwebp`, add `pixman` as makedep

ImperatorStorm commented:
Also adding on `pixman` as a makedep, as we're using the shared library of it in aseprite.

ImperatorStorm commented:
~~Ok `ld` failed to link, not sure what's going on there.~~

~~`ld` wants WebPDemux and WebPMux stuff, though it should have been built.~~

> gif2webp or img2webp is required to get `libwebpdemux`

Hmm, `-DWEBP_BUILD_GIF2WEBP=YES` should be building it, not sure why `ld`'s complaining.

> 2022-01-05T07:02:55.3726080Z CMake Warning:
2022-01-05T07:02:55.3727021Z   Manually-specified variables were not used by the project:
...
2022-01-05T07:02:55.3729311Z     WEBP_BUILD_GIF2WEBP
2022-01-05T07:02:55.3729704Z     WEBP_BUILD_IMG2WEBP
...

Odd, looks like shared `libwebp` breaks this.

ImperatorStorm committed:
possibly fix shared-libwebp.patch?

ISSOtm commented:
That'd be normal, as `libwebpmux` and `libwebpdemux` are also part of `libwebp`.

ISSOtm committed:
Use system HarfBuzz and FreeType

I remember getting build errors, but cannot reproduce them anymore.
So use the system libs more where possible!

ISSOtm committed:
Find all WebP libs

ISSOtm committed:
Mark libwebp as runtime dep

ISSOtm committed:
Remove CMake variables used by libwebp build

They do nothing now that we are using shared libwebp

ISSOtm commented:
Seems like this works just fine, and I also managed to make `harfbuzz` and `freetype` shared :)

ISSOtm commented:
> Also adding on `pixman` as a makedep, as we're using the shared library of it in aseprite.

The shared lib doesn't appear to be *used*, since the LAF backend is Skia. I'll check if we can skip having it installed at all.

ImperatorStorm commented:
> Should we require clang? Skia's build instructions mention that performance may be significantly worse with non-Clang compilers

Regarding this, it would likely be best to require clang.

ISSOtm commented:
Thing is, I don't know if it really holds true anymore. But I don't know how we'd bench performance either.

ImperatorStorm commented:
Fair, might be best to not force clang then.

ISSOtm commented:
Ah, I see that Skia's source has a `bench` subdirectory. Gimme a while to look at that.

ISSOtm commented:
Looking more into it, it doesn't seem like it would make sense to bench Skia itself, but we'd be more interested in how Aseprite uses it. So I don't know. Maybe we should stick to the default for now (which can be configured in `makepkg.conf`, mind you), and pin a message in the AUR comments telling users to report any performance issues. Based on feedback we can then choose whether to require Clang.

ImperatorStorm commented:
sounds good.

ImperatorStorm commented:
> Do we really conflict with `skia-git`? If so, that should be reinstated in the conflicts array

I can't imagine why we'd conflict with `skia-git`, probably shouldn't re-add it to `conflicts`.

ImperatorStorm commented:
> Run tests? E.g. LAF's, except that they fail to build due to an obscure linker error (looks like some static lib requires a dynamic lib that doesn't get -l'd?)

Should probs give it a try.

ISSOtm commented:
No, we shouldn't `conflicts` with them. *They* should `provides` us, and also `conflicts` with us.

ISSOtm commented:
> > Do we really conflict with `skia-git`? If so, that should be reinstated in the conflicts array
> 
> I can't imagine why we'd conflict with `skia-git`, probably shouldn't re-add it to `conflicts`.

I was thinking there could be a lib search path conflict, but I really don't think it can happen, so this is very likely the right move imo.

ISSOtm commented:
The last thing I want to check is compiling less things into Skia, to reduce build time. My plan is to remove libs and see if compilation fails. AIUI, since we are *statically* linking to Skia, if a function relied on by Aseprite is not present, we should get a build-time error. (Right?)

ISSOtm committed:
Remove build-time dep on Pixman

It doesn't wind up being used at all in the end

ISSOtm committed:
Trim off a bunch of dependencies from Skia

Only libpng is required to render some of the images, it seems,
so this reduces build time and the amount of sources, which is good!

ISSOtm commented:
Bruh, I'm stupid.

ISSOtm commented:
Sorry about that, I added the missing file. My bad!

ImperatorStorm commented:
> > Run tests? E.g. LAF's, except that they fail to build due to an obscure linker error (looks like some static lib requires a dynamic lib that doesn't get -l'd?)
> 
> Should probs give it a try.

wait on this or merge?

ISSOtm commented:
Good enough imo. Let's not wait indefinitely on the changes.

ImperatorStorm commented:
Manually merged into [`master`](https://github.com/ImperatorStorm/PKGBUILDs/tree/master/).

Thanks for rewriting the package!

ISSOtm commented:
Thank you for collaborating as well! :)

ImperatorStorm commented:
squash+merged into `aseprite` because AUR got mad when i tried to push aseprite.
